Zotero/Scenes/General/ViewModels/NoteEditorActionHandler.swift

```diff
@@ -8,29 +8,122 @@
+        self.dbStorage = dbStorage
+        self.schemaController = schemaController
+        self.saveCallback = saveCallback
+        backgroundQueue = DispatchQueue(label: "org.zotero.Zotero.NoteEditorActionHandler.queue", qos: .userInteractive)
+    }
 
     func process(action: Action, in viewModel: ViewModel<NoteEditorActionHandler>) {
         switch action {
         case .save:
-            self.saveAction(viewModel.state.text, viewModel.state.tags)
+            save(viewModel: viewModel)
 
         case .setText(let text):
             guard text != viewModel.state.text else { return }
-            self.update(viewModel: viewModel) { state in
+            update(viewModel: viewModel) { state in
                 state.text = text
                 state.changes = .save
             }
 
         case .setTags(let tags):
-            self.update(viewModel: viewModel) { state in
+            update(viewModel: viewModel) { state in
                 state.tags = tags
                 state.changes = [.tags, .save]
             }
         }
+
+        func save(viewModel: ViewModel<NoteEditorActionHandler>) {
+            let kind = viewModel.state.kind
+            let library = viewModel.state.library
+            let text = viewModel.state.text
+            let tags = viewModel.state.tags
+
+            switch kind {
+            case .itemCreation(let parentKey):
+                createItemNote(library: library, parentKey: parentKey, text: text, tags: tags)
+
+            case .standaloneCreation(let collection):
+                createStandaloneNote(library: library, collection: collection, text: text, tags: tags)
+
+            case .edit(let key):
+                updateExistingNote(library: library, key: key, text: text, tags: tags)
+
+            case .readOnly:
+                let error = State.Error.cantSaveReadonlyNote
+                DDLogError("NoteEditorActionHandler: can't update read only note: \(error)")
+                saveCallback(.failure(error))
+            }
+
+            func createItemNote(library: Library, parentKey: String, text: String, tags: [Tag]) {
+                let key = KeyGenerator.newKey
+                let note = Note(key: key, text: text, tags: tags)
+                let type = schemaController.localized(itemType: ItemTypes.note) ?? ItemTypes.note
+                let request = CreateNoteDbRequest(note: note, localizedType: type, libraryId: library.identifier, collectionKey: nil, parentKey: parentKey)
+                createNote(note, with: request)
+            }
+
+            func createStandaloneNote(library: Library, collection: Collection, text: String, tags: [Tag]) {
+                let key = KeyGenerator.newKey
+                let note = Note(key: key, text: text, tags: tags)
+                let type = schemaController.localized(itemType: ItemTypes.note) ?? ItemTypes.note
+                let collectionKey: String?
```

Comment:
There are some helper extensions for `Collection`, so you could probably do just: `let collectionKey = collection.isCollection ? collection.identifier.key : nil`

Reply:
Check https://github.com/zotero/zotero-ios/pull/826/commits/01c985e8f05f204526ffaba59aa4719e745f2f88